### PR TITLE
[4.0] System - HTTP Headers - Write headers to the configuration file

### PR DESCRIPTION
**Pull request by zero-24**
### Summary of Changes

- Adds the new option `Write headers to the configuration file` (source https://github.com/zero-24/plg_system_httpheader/pull/6)
- restructure the plugin
- cleaup xml, doc blocks and php code

#### Test  `Write headers to the configuration file`
- Install Joomla 4.0
- apply this patch
- go to the plugin config
- enable the new option `Write headers to the configuration file`
- save the plugin
- check the content of the .htaccess | web.config file
- change the static header configurations
- save the plugin
- check the .htaccess | web.config file

#### Test  the CSP Component Integration

- go to com_csp
- enable the detect option
- visit the backend and frontend
- notice the collected reports
- publish some of the reports
- enable the auto mode from the options (keep it report only for now :D)
- check the now generated csp header
- enable the custom mode
- configure a custom rule (adding script-src 'nonce-{nonce}' as minimum)
- check that `{nonce}` has been replaced by the actual nonce

### Expected result

The static headers are written to the respective server config file

### Actual result

This option does not exists

### Documentation Changes Required

- [ ] Update Screenshots (Plugin and Component options)
- [ ] Add new option description

https://docs.joomla.org/J4.x:Http_Header_Management

### CODE REVIEW

**Comment by zero-24** [comment]
@wilsonge I would like to add yet another feature that sets the `frame-ancestors` directive when enabled can you give me a status how do you want to handle this PR?

**Comment by zero-24** [comment]
And I have been asked by @SniperSister to work on an approach for hash based whitlisting too, wich would massive conflict with this PR here too.

**Comment by zero-24** [comment]
Any update @wilsonge ?

File: plugins/system/httpheaders/httpheaders.php [hunk 1]
```diff
@@ -124,119 +167,113 @@ public static function getSubscribedEvents(): array
...
+		$cspMode = $this->comCspParams->get('contentsecuritypolicy_mode', 'detect');
 
 		// In detecting mode we set this default rule so any report gets collected by com_csp
 		if ($cspMode === 'detect')
 		{
 			$frontendUrl = str_replace('/administrator', '', Uri::base());
 
 			$this->app->setHeader(
-				'Content-Security-Policy-Report-Only',
+				'content-security-policy-report-only',
```
Comment (wilsonge) [inline plugins/system/httpheaders/httpheaders.php:276]: Don't we want to set the headers in the app with the correct cases?

Comment (zero-24) [reply plugins/system/httpheaders/httpheaders.php:276]: > Each header field consists of a name followed by a colon (":") and the field value. Field names are case-**in**sensitive. 

https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2

I personally would go for the the lowercase version as it is now.

**Comment by wilsonge** [comment]
Thanks!

**Comment by zero-24** [comment]
Great, thanks 🎉

**Comment by brianteeman** [comment]
This should not have been merged https://github.com/joomla/joomla-cms/pull/25713#issuecomment-515661783

**Comment by zero-24** [comment]
Removing the documentation required label as this feature is about to be removed here: https://github.com/joomla/joomla-cms/pull/25754

**Pull request merged**